### PR TITLE
merge DeepEl back into develop

### PR DESCRIPTION
**Pull request by 0UmfHxcvx5J7JoaOhFSs5mncnisTJJ6q**

### CODE REVIEW

**Comment by merfort** [comment]
Looks good from my side.

**Review by fschreyer — COMMENTED** [review]
Cool! :) Merge seemed to be quite smooth after all. I see no issues from my side.

**Review by Renato-Rodrigues — CHANGES_REQUESTED** [review]
1. There are some comment duplications unnecessary added. Ex:
```
# cfg$gms$cm_flex_tax: switch to switch flexibility tax on or off
# cfg$gms$cm_PriceDurSlope_elh2:
... and a few others 

```

2. Are the change on core/bounds for biotr changing the default remind behavior? Is this intentional?

3. Please revert the change for the year in the file: core/magicc/read_DAT_SURFACE_TEMP.R 

4. Is it intentional the removal of the initialization values for the switches below in the main.gms?
```
cm_indst_H2costAddH2Inv = 0.1;  !! def 6.5$/kg = 0.2 $/Kwh
cm_indst_costDecayStart = 0.05; !! def 5%
cm_indst_H2costDecayEnd = 0.1;  !! def 10%
```
or any other switches removal in this file?

**Comment by 0UmfHxcvx5J7JoaOhFSs5mncnisTJJ6q** [comment]
> 1. Are the change on core/bounds for biotr changing the default remind behavior? Is this intentional?

No, since 
```
cfg$gms$cm_tradbio_phaseout <- "default" #def <- "default"
```

> 2. Please revert the change for the year in the file: core/magicc/read_DAT_SURFACE_TEMP.R

Those files are gone anyhow (704f333)

> 3. Is it intentional the removal of the initialization values for the switches below in the main.gms?
> 
> ```
> cm_indst_H2costAddH2Inv = 0.1;  !! def 6.5$/kg = 0.2 $/Kwh
> cm_indst_costDecayStart = 0.05; !! def 5%
> cm_indst_H2costDecayEnd = 0.1;  !! def 10%
> ```

No, as it leads to compilation errors.  Pull request into MariannaR:develop_DeepEl is [pending](https://github.com/MariannaR/remind/pull/3).

> or any other switches removal in this file?

Are there more?  If so, `testOneRegi` didn't notice.

File: core/datainput.gms
Comment (Renato-Rodrigues) [reply core/datainput.gms:149]: What was the change exactly to the generisdata_tech file?
If it is a scenario assumption, why not use the switches that exist to modify the technology parameters instead of directly overwriting the original file?

Comment (silviamade) [reply core/datainput.gms:149]: I can comment on the changes in the generisdata_tech file. Changes were on inco0, incolearn, learn for storspv and learn for spv. I can provide old vs new values if needed

Comment (0UmfHxcvx5J7JoaOhFSs5mncnisTJJ6q) [reply core/datainput.gms:149]: `generisdata_tech.prn` is unchanged.  The changes I referred to are lines 149 pp., but for some reason Github shows the three lines before the one where I inserted the comment.
There's a new switch (`cm_VRE_supply_assumptions`), but the data it is swapping out was provided by Falko through e-mail/Mattermost, and we should have a more reliable way of figuring out where that data came from (i.e. `mrremind`).

Comment (Renato-Rodrigues) [reply core/datainput.gms:149]: just to be more clear of what I meant. 

I introduced a series of switches that modify technologies assumptions during the innopaths project in order to reflect different scenarios. ex: cm_INNOPATHS_adj_coeff_multiplier, cm_INNOPATHS_adj_seed, cm_INNOPATHS_adj_coeff, cm_INNOPATHS_incolearn, ....

You can simply use them directly in the scenario config file for the scenario you are creating instead of hard-coding your specific scenario situation to the remind code.

This is preferable because if we all start hard-coding multiple paper and project specific scenarios in the remind code, the code will become very lengthy and a mess. It will also be very difficult to figure out where in the code are all relevant switches necessary to reflect a specific scenario assumption.

Instead, if you use the switches, we will always be able to check the DeepEl scenario config file directly and have all information on how each scenario is defined, and which switches play a role on it. This is already done in the innopaths and ariadne project.

PS: the INNOPATHS keyword will be removed from these switches in the near future.

Comment (Renato-Rodrigues) [reply core/datainput.gms:149]: > I can comment on the changes in the generisdata_tech file. Changes were on inco0, incolearn, learn for storspv and learn for spv. I can provide old vs new values if needed

Thank you Silvia.

If you are changing the generisdata_tech file directly, you are affecting every other remind run. We would need to check the motivation behind those changes and all values that changed to decide if we assume those numbers as our new default behavior.

Ideally you should use the switches I mentioned above to avoid that.

Comment (Renato-Rodrigues) [reply core/datainput.gms:149]: > `generisdata_tech.prn` is unchanged. The changes I referred to are lines 149 pp., but for some reason Github shows the three lines before the one where I inserted the comment.
> There's a new switch (`cm_VRE_supply_assumptions`), but the data it is swapping out was provided by Falko through e-mail/Mattermost, and we should have a more reliable way of figuring out where that data came from (i.e. `mrremind`).

Thank you Michaja. 

Then I would second the request to use switches instead. To me it makes no sense to add a multitude of alternative scenarios to the remind code, increasing significatively its length if you can define then in a scenario config file.

PS: sorry for replying out of order, I saw your messages in the wrong order.

Comment (Renato-Rodrigues) [reply core/datainput.gms:149]: For example: 
- for incolearn you can simply set the switch to cm_INNOPATHS_incolearn "storspv 4240" in your scenario config.
- for inco0, you can either use the cm_INNOPATHS_inco0Factor switch or create a similar switch that simply overwrites the default value following the same principle of cm_INNOPATHS_incolearn, or  cm_INNOPATHS_adj_coeff
- for learn, I am not sure if we already have a switch in place, but in case not you can do the same as above.

Comment (MariannaR) [reply core/datainput.gms:149]: @0UmfHxcvx5J7JoaOhFSs5mncnisTJJ6q , I agree with you that committing a tailor-made `generisdata` file should be avoided. In my opinion we should not include the new VRE assumptions as they would require additional validation and a `mrremind` function behind.

> 
> 
> This would be the moment to discuss whether this can be put on a more permanent, less ad-hoc basis.
> VRE supply assumptions where by @FalkoUeckerdt, iirc. Any thoughts?

Comment (0UmfHxcvx5J7JoaOhFSs5mncnisTJJ6q) [reply core/datainput.gms:149]: @Renato-Rodrigues 
https://github.com/remindmodel/remind/blob/cca768efdc0e5276da5525ab18d3610a57b8b329/core/datainput.gms#L176-L179
Clever.  :)

Problem is, `cm_VRE_supply_assumptions` also affects equations: [`q32_limitCapTeStor`](https://github.com/remindmodel/remind/pull/395/files/1dfb113cdb2f9ecf0e1bcb5db78812a20ab85dbb#diff-e1607788f31836706fabc5462a482129fd8990bce395c5a62ef59c2418c7d899R52-R65) and [`q32_shStor`](https://github.com/remindmodel/remind/pull/395/files/1dfb113cdb2f9ecf0e1bcb5db78812a20ab85dbb#diff-e1607788f31836706fabc5462a482129fd8990bce395c5a62ef59c2418c7d899R132-R145), so I think the cleanest way is to keep the switch and put the modified parameters into `mrremind`.

**Comment by robertpietzcker** [comment]
I started to write
"I would be in favor of having them implemented as a switch in REMIND. I find it quite likely that in the next 2-3 years there will be requests like "let's look how this compares to a new scenario with the DeepEl VRE changes".  "

but then I thought "the VRE costs & integration costs will likely change over the next years, so having a switch that say "storspv cost *0.5" will have a different meaning in a year compared to today.

so probably it would make most sense to have an (internal) documentation on redmine or so, where it is clearly written down which parameters (inco0, learn, p32_factorstorage, ...) were a) set to which value, and b) how this is different compared to the default value in this REMIND version.

then a person in a year can look at the then-current values of parameters and use existing switches to scale up/down the parameters to be in line with the DeepEl scenarios
  
Best,
Robert

File: core/datainput.gms
Comment (robertpietzcker) [reply core/datainput.gms:149]: I started to write
"I would be in favor of having them implemented as a switch in REMIND. I find it quite likely that in the next 2-3 years there will be requests like "let's look how this compares to a new scenario with the DeepEl VRE changes". "

but then I thought "the VRE costs & integration costs will likely change over the next years, so having a switch that say "storspv cost *0.5" will have a different meaning in a year compared to today.

so probably it would make most sense to have an (internal) documentation on redmine or so, where it is clearly written down which parameters (inco0, learn, p32_factorstorage, ...) were a) set to which value, and b) how this is different compared to the default value in this REMIND version.

then a person in a year can look at the then-current values of parameters and use existing switches to scale up/down the parameters to be in line with the DeepEl scenarios

Best,
Robert

Comment (Renato-Rodrigues) [reply core/datainput.gms:149]: I basically follow two principles:

(1) I keep any switch implemented in remind as close as possible to the original data assumption.
In that sense I avoid using the "multiplying" switches (ex: cm_INNOPATHS_adj_coeff_multiplier) and try to favor absolute value ones (ex: cm_INNOPATHS_adj_coeff) whenever possible. This reduces the problem of changes in the input data affecting the scenario assumptions, so in the future you can rerun the paper/project scenario even if the model changed in between.

(2) For anything that demands a group of assumption changes, I implement a switch at the most primitive level as possible to represent each of the components that compose the scenario. I then only really define the scenario by configuring these switches in the scenario config file. 
This allows us much easier re-use of switches created for other projects/papers, and also avoids the existent of switches that modify multiple assumptions at the same time, which are much less customizable and become obsolete much faster.

@0UmfHxcvx5J7JoaOhFSs5mncnisTJJ6q I would still define the scenario using the scenario config file instead, and just create additional switches that enable/disable the changes in the q32_limitCapTeStor and q32_shStor equations to reflect the additional assumptions of the VRE scenario.

Comment (Renato-Rodrigues) [reply core/datainput.gms:149]: On the documentation part, if we follow the principle of adding only "primitive" switches to the remind code, their documentation would belong to the default cfg file as it is already done.

For any scenario specific documentation, you can easily add description lines to your scenario config csv file like I already do for sub-groups of REMIND-EU scenarios.
Just add a line describing the scenario before it, and disable it.

Comment (0UmfHxcvx5J7JoaOhFSs5mncnisTJJ6q) [reply core/datainput.gms:149]: There's a trade-off between flexibility (tuning specific parameters using individual switches) and legibility (this switch does the thing it is named for).  Having to modify four switches in a specific way in order to achieve one scenario setting is quite error-prone (speaking from experience).
I don't much care either way, and I'm not sure how to reach a consensus here.  We might merge as-is or split for multiple switches and create an issue to review this and clean it up later on, once we have word from @FalkoUeckerdt what the fate of this parameters is.

Comment (robertpietzcker) [reply core/datainput.gms:149]: I disagree about it being sufficient to document this in the scenario_config. 

1-2 years later when the questions come about what the specific VRE assumptions were in paper xyz, just having the scenario_config will not give you a good answer, because you first have to dig up an old remind version and then spend some hours trying to understand how the many different switches interacted and which is actually relevant for what you are trying to find out. 
I think it is much more accessible if you write it down in a fix place that can be easily found (eg REMIND wiki, if somebody remembers to search there). 

Best,
Robert
> 
> On the documentation part, if we follow the principle of adding only "primitive" switches to the remind code, their documentation would belong to the default cfg file as it is already done.
> 
> For any scenario specific documentation, you can easily add description lines to your scenario config csv file like I already do for sub-groups of REMIND-EU scenarios.
> Just add a line describing the scenario before it, and disable the line so its not considered while creating new runs.

Comment (Renato-Rodrigues) [reply core/datainput.gms:149]: I would argue that besides the paper, it is much easier to find the information about a scenario in the file that you need to open anyway to run the scenario than in a two years old wiki page somewhere.

The main principle of coding documentation is to add it as close as possible to the related code. To me, what defines a scenario in remind are the options you define to remind run using the scenario config file, therefore my logic on either adding the documentation to the scenario config file directly, or alternatively, create a metadata file that you need to commit together with any scenario config that described the scenarios contained on it.

**Review by robertpietzcker — COMMENTED** [review]
I strongly disagree!

I don't care if the docu is also in the scenario_config - actually it is great if it also there - but a scenario_config really is NOT a good format for easy readibility and understandability.
 
Having a text with some subheaders for the main aspects is MUCH easier to read and understand later. 

I really don't understand what is the downside of simply typing up 1-2 pages of docu for the main changes and uploading them to the wiki? There is a high chance it will reduce workload for the people who have to try and find this information sometime in the future, and you can always add the sentence "please look in the scenario_config_xxxx that should be stored here: yyy for more info". 

so please, do me the favor and ALSO do some documentation of the main changes for deepEl in somewhat structured bullet point list and put it somewhere easily findable - I would propose the remind wiki, because a) it is PIK-internal, b) it can be accessed by all people in the REMIND group, c) searching works ok, d) it won't go away in the next years

the paper has some info, but is always geared towards an external person, so it will not use the REMIND-specific parameter naming etc. 



(and I have no idea why this has turned from a reaction on Renato's last text into a "review" - I can't find a button to stop it, so I'll just "submit review" with this comment)

File: core/datainput.gms
Comment (0UmfHxcvx5J7JoaOhFSs5mncnisTJJ6q) [reply core/datainput.gms:149]: > the remind wiki […] won't go away in the next years

The Wiki on Redmine is really ephemeral.  There used to be a lot of information, that isn't there anymore.  Some of it got moved to the `./tutorials/` directory, some of it didn't.  In any case, any links that existed to the wiki are dead.
I'm all for verbose documentation in plainly written English (as opposed to some GAMS code compressed into a .csv column).  And I also think it should stick to the code like glue (i.e. included with the repository), else it will get lost.

Just my two cents.

Comment (Renato-Rodrigues) [reply core/datainput.gms:149]: > Having a text with some subheaders for the main aspects is MUCH easier to read and understand later.
> 

If you read my message again, you will see that I mentioned adding a metadata file to document the scenario config.
This file does not need, and shouldn't, follow a csv file format. Instead I would use a markdown document format as it is already used by other documentation shipped with the model.

And, although I respect your opinion, I am strongly in favor of shipping this documentation file together with the scenario config, as it allows easy access, updating and versioning.

Comment (cchrisgong) [reply core/datainput.gms:149]: If I may add because I am also thinking about how to use the tech switches: how do these three tech assumptions compare to each other?
1) c_techAssumptScen
2) cm_VRE_supply_assumptions
3) Falko's data in generisdata_tech.prn

Do they overlap? If yes, should we maybe bring them under one hood at some point so they are all together in one place? Otherwise we might accidentally double count "optimism" by turning them all on!

**Comment by silviamade** [comment]
Dear all, I think we should reach a consensus here otherwise this discussion might never be closed. Since the merge will affect all users and it seems that different users prefer different solutions, I think we should maybe have a meeting to discuss the issue involving also Lavinia and Falko (and Gunnar if he has time)? Otherwise, do you have other ideas to avoid a meeting and take a final decision? Thanks!

File: core/datainput.gms
Comment (0UmfHxcvx5J7JoaOhFSs5mncnisTJJ6q) [reply core/datainput.gms:149]: > 1. c_techAssumptScen

https://github.com/remindmodel/remind/blob/cca768efdc0e5276da5525ab18d3610a57b8b329/config/default.cfg#L792-L796

> 2. cm_VRE_supply_assumptions

https://github.com/MariannaR/remind/blob/4e20873024e4466c4b9f2e0051b3535f8951a8ef/config/default.cfg#L946-L954
```
# cfg$gms$cm_VRE_supply_assumptions: 
#   for 1 - optimistic, modify
#     - inco0, incolearn, and learn parameters for spv and storspv
#     - ease capacity constraints on storage
#     - reduce necessary storage for electricity production
#  for 2 - sombre, modify
#     - incolearn spv to 5010 (150 $/kW floor cost)
#  for 3 - bleak, modify
#     - incolearn spv to 4960 (200 $/kW floor cost)
```
> 3. Falko's data in generisdata_tech.prn

If you are referring to the data described in [Falko's Mattermost post](https://mattermost.pik-potsdam.de/rd3/pl/cyb16pjze3f95br7ue93icrw5e)
> The solar PV CAPEX and learning rate were reduced representing optimistic ranges in here https://onlinelibrary.wiley.com/doi/full/10.1002/pip.3189
> The battery costs and learning rate are in accordance with https://www.nature.com/articles/nenergy2017110
> On the storage requirement reduction with electrification it is:
> https://arxiv.org/abs/1801.05290  and
> https://www.sciencedirect.com/science/article/pii/S2542435120303408

that data is _not_ in `generisdata_tech.prn`, but being spliced in by `cm_VRE_supply_assumptions`.

> Do they overlap? 

Yes.

> If yes, should we maybe bring them under one hood at some point so they are all together in one place? Otherwise we might accidentally double count "optimism" by turning them all on!

No.  `cm_VRE_supply_assumptions` are overwritten by `c_techAssumptScen`.  But this could sure use some clarification.

I don't see us solving this here in this issue, since the discussion already devolved into an argument about the general policy for implementing switches in REMIND.  I suggest to implement this as is, since @FalkoUeckerdt  didn't indicate this to be implemented in the inputdata anytime soon and work on `DeepEl` has to continue.
Probably there should be a general discussion on switch-philosophy in the REMIND meeting.  This can be adopted to whatever is decided there.

**Review by Renato-Rodrigues — COMMENTED** [review]
I don't think I have any other comments in the merge. 

Nevertheless, I am still of the opinion that is better and very straightforward to use already existent switches to change the inco0 and incolearn values instead of adding 37 extra lines of code to the datainput file.

**Review by Renato-Rodrigues — APPROVED** [review]

**Comment by silviamade** [comment]
Ok Marianna, is doing test runs, let's wait for the results before we complete the merge. I second the suggestion of Michaja to have a general discussion on switch-philosophy in the REMIND meeting

**Comment by 0UmfHxcvx5J7JoaOhFSs5mncnisTJJ6q** [comment]
> I second the suggestion of Michaja to have a general discussion on switch-philosophy in the REMIND meeting

I put it down [as a topic](https://redmine.pik-potsdam.de/projects/remind-r/wiki/REMIND_meeting_minutes#362021).  @Renato-Rodrigues, @robertpietzcker maybe you want to prepare some pro and other-pro arguments to facilitate the discussion.

File: core/datainput.gms
Comment (MariannaR) [reply core/datainput.gms:149]: does anyone know if this switch `c_H2tdCapCost_stat` has to be kept, and where does it belong? maybe @Renato-Rodrigues and @rkrekeler and @fschreyer ?

Comment (Renato-Rodrigues) [reply core/datainput.gms:149]: felix would know exactly, but I believe he removed this switch in one of his recent commits that include the latest updates to the ariadne changes:
https://github.com/remindmodel/remind/commit/dbbfdec3adb7e655b0b2146eecd0c72abfd2c515

Comment (fschreyer) [reply core/datainput.gms:149]: can be removed

**Review by merfort — APPROVED** [review]

**Pull request merged**
